### ISSUE TITLE
[Problem]: dbus metadata was not getting updated for radio, YouTube etc 

### ISSUE THREAD

**Issue by phelbas** [mikebrady/shairport-sync#1560]
### What happened?

using shairport-display-qt.py I was getting nothing displayed at all for directing stuff to airplay2 without cover art, or other meta data. 

Using busctl monitor org.gnome.ShairportSync I did not see metadata getting sent out for those streams.

I fixed it by making  int metadata_packet_item_changed = 0;  in metadata_hub.c
into static int

I believe this function processes one packet at a time? it only uses metadata_packet_item_changed  (passing it out via metadata_hub_modify_epilog) when it sees an mden packet so I didn't quite understand the purpose of a local variable being lost over multiple packets. Now it is set to zero when mdst is seen and set to 1 on any meta data update packet, then used on mden packet.

Anyway changing it to static fixed the issue for me. Now I see dbus metadata when switching between streams like radio -> YouTube -> whatever -> music
this might be a stupid fix?

Also because streams do not necessarily contain all the metadata such as cover art, I modified shairport-display-qt to be cope with missing metadata keys as currently this code generates a quiet exception if any metadata is missing. I thought this was the reason I wasn't getting displays on YouTube etc but actually busctl monitor  showed no metadata was appearing..

### Relevant log output

```shell
here is what a YouTube metadata output looks like on dbus now:

STRING "Metadata";
                          VARIANT "a{sv}" {
                                  ARRAY "{sv}" {
                                          DICT_ENTRY "sv" {
                                                  STRING "mpris:trackid";
                                                  VARIANT "o" {
                                                          OBJECT_PATH "/org/gnome/ShairportSync/FEFCE8";
                                                  };
                                          };
                                          DICT_ENTRY "sv" {
                                                  STRING "xesam:title";
                                                  VARIANT "s" {
                                                          STRING "Fresh & Ready To Go! // VW Beetle 1600 Dual Port";
                                                  };
                                          };
                                          DICT_ENTRY "sv" {
                                                  STRING "xesam:artist";
                                                  VARIANT "as" {
                                                          ARRAY "s" {
                                                                  STRING "Sarah -n- Tuned";
                                                          };
                                                  };
                                          };
                                          DICT_ENTRY "sv" {
                                                  STRING "mpris:length";
                                                  VARIANT "x" {
                                                          INT64 942401000;
                                                  };
                                          };
                                  };
                          };
```

### Configuration Information.

4.1-dev-708-ge7c6c4b2-AirPlay2-alac-OpenSSL-Avahi-ALSA-pipe-soxr-metadata-dbus-sysconfdir:/etc

### How did you install Shairport Sync?

Built from source

### Check previous issues

- [X] Confirm

**Comment by mikebrady** [comment]
Thanks for the very interesting report! Let me check it out, but it’s entirely possible your fix is a good one.

**Comment by mikebrady** [comment]
Taking a quick look at it, I think you are correct – there does seem to be a fault in the logic, and making `metadata_packet_item_changed` fixes it. Let me think it through. Many thanks for finding this problem!

**Comment by mikebrady** [comment]
So an update has been pushed to `development` with basically your fix in it. If you got a chance to try it out and report back, that would be great.

**Comment by phelbas** [comment]
thanks for the fast response!

**Comment by mikebrady** [comment]
@phelbas, have you published your version of `shairport-display-qt` anywhere?

**Comment by phelbas** [comment]
have been meaning to create a pull request will get onto it.

**Comment by phelbas** [comment]
@mikebrady here is the first version of the forked shairport-display-qt https://github.com/phelbas/shairport-display-qt

**Comment by mikebrady** [comment]
> @mikebrady here is the first version of the forked shairport-display-qt https://github.com/phelbas/shairport-display-qt

Very nice it is too. I'm playing with it right now, and might have a few simple suggestions.

**Comment by phelbas** [comment]
@mikebrady hey Mike I'm hacking around with it offline to see if I can make it more IOS like with background gradient picked from album art and so on but struck a question. Using dbus, how do I query the current player state? for example, from *phone*, if I pause/play I don't see any dbus signals sent. And at start up I am not sure how to query for whether shairport-sync is playing, or paused and where it is in the song?

**Comment by mikebrady** [comment]
Okay, there are two issues here:
1. How do you find out what capabilities are in the Shairport Sync D-Bus interface?
2. How do you access them?

For (1), I suggest you use `dfeet`, a GUI tool for exploring D-Bus interfaces on Linux. Here is a screenshot of it looking at Shairport Sync (on Fedora 36):

<img width="879" alt="image" src="https://user-images.githubusercontent.com/4265913/200797580-2dcbf618-1d9c-44f8-9594-272975a09bdd.png">


For (2), I copied some of the code that is already in `shairport-display-qt` to make up a simple method to read items from the Shairport Sync D-Bus service. It's very crude, but hopefully it will show you how to go about it. Here is the method, and I put it in after the definition of the method `_set_backlight`:
```python
    def _get_sps_info(self, path, item):
        try:
            result = self._bus.call_blocking(
                "org.gnome.ShairportSync",
                "/org/gnome/ShairportSync",
                "org.freedesktop.DBus.Properties",
                "Get",
                "ss",
                ["org.gnome.ShairportSync" + path, item],
            )
        except dbus.exceptions.DBusException:
            self.log.warning("_get_sps_info failed")
            return None
        return result
```

The idea is that the `path` argument (if any) is added to `org.gnome.ShairportSync` to form the full interface pathname, e.g. `org.gnome.ShairportSync.RemoteControl` and the `item` is the name of the property you want to get the value of, e.g. `PlayerState`. Here is an example of a few calls, stuck in just before the display is initialised:
```python
        # Existing code...
        self.Elapsed.setFont(QFont("Montserrat", 10, QFont.Normal))

        # Sample code... 
        print(self._get_sps_info("", "Protocol"))
        if self._get_sps_info(".RemoteControl", "Available") != 0:
            print(self._get_sps_info(".RemoteControl", "Volume"))
            print(self._get_sps_info(".RemoteControl", "ProgressString"))
            print(self._get_sps_info(".RemoteControl", "PlayerState"))
        else:
            print("Remote control is not available")

        # Existing code...
        self._clear_display()
```

**Comment by mikebrady** [comment]
> For example, from iPhone, if I pause/play I don't see any dbus signals sent. 

Whenever a property changes in the Shairport Sync D-Bus interface, `shairport-display-qt` is set up to receive a `PropertiesChanged` signal, but the method it calls in response, `self._display_metadata`, only looks at the metadata. Here is the code that sets all that up:
```python
    def _setup_signals(self):

        self.properties_changed = self._bus.add_signal_receiver(
            handler_function=self._display_metadata,
            signal_name="PropertiesChanged",
            dbus_interface="org.freedesktop.DBus.Properties",
            bus_name="org.gnome.ShairportSync",
            member_keyword="signal",
        )
```

So, I'm pretty sure that changes such as `play` and `pause` and so on do indeed send a `PropertiesChanged` signal, but the program is only reading metadata and not at any of the other properties...

**Comment by phelbas** [comment]
Thanks. well I am using this to monitor:
```
sudo dbus-monitor --system
```

and it shows everything, I think. But nothing happens on the bus when I press Pause on the phone, even though music stops. When I press play, I get dbus meta data with extra fields.

In a half second timer I used your example code to query the RemoteControl, and get 

<img width="525" alt="Screen Shot 2022-11-10 at 12 18 22 am" src="https://user-images.githubusercontent.com/67846089/200840615-b2a451e5-0660-4399-9b6e-3cabca5074e3.png">

```
4119954867/4121315850/4124920879
Playing
4119954867/4121315850/4124920879
Playing
4119954867/4121315850/4124920879
Playing
4119954867/4121315850/4124920879
Playing
4119954867/4121315850/4124920879
```

whether or not I've paused using phone or using interface to Pause. (Playing doesn't change) 
Note also the ProgressString and PlayerState returned by self._get_sps_info isn't changing. (The original shairport-sync-qt does its own countdown timing starting at the meta data change).

Fixing the ability to query the state will allow me to push this update which picks a gradient from the cover art :)

<img width="797" alt="Screen Shot 2022-11-10 at 12 20 12 am" src="https://user-images.githubusercontent.com/67846089/200841012-92d6a511-07b2-4e3a-a807-52abee21ad43.png">

**Comment by mikebrady** [comment]
Thanks. Let me dig in to this.

**Comment by mikebrady** [comment]
Looks nice, BTW!

**Comment by mikebrady** [comment]
It might be worth updating to the latest version of Shairport Sync on the `development` branch, as the treatment of pause on a buffered audio stream (e.g. from the iOS Music app) has been dealt with -- it might account for the missing `Paused` messages.

On the latest, it seems to be working. This is pausing a Music track:
```
signal time=1668003364.784260 sender=:1.384 -> destination=(null destination) serial=49 path=/org/gnome/ShairportSync; interface=org.freedesktop.DBus.Properties; member=PropertiesChanged
   string "org.gnome.ShairportSync.RemoteControl"
   array [
      dict entry(
         string "PlayerState"
         variant             string "Paused"
      )
   ]
   array [
   ]
method return time=1668003364.784999 sender=:1.384 -> destination=:1.369 serial=50 reply_serial=395
```
Here's output from the app itself:
```
2022-11-09 14:38:42 - [DEBUG] - elapsed: 0:01:22
2022-11-09 14:38:42 - [DEBUG] - starting timer
2293466282/2297083277/2301500586
Playing
2293466282/2297083277/2301500586
Playing
2293466282/2297083277/2301500586
Playing
2293466282/2297083277/2301500586
Playing
2293466282/2297083277/2301500586
Playing
2293466282/2297083277/2301500586
Playing
2022-11-09 14:38:46 - [DEBUG] - Received signal for org.gnome.ShairportSync.RemoteControl
2022-11-09 14:38:46 - [DEBUG] - starting timer
2022-11-09 14:38:46 - [INFO] - PlayerState: Paused
2293466282/2297083277/2301500586
Paused
2293466282/2297083277/2301500586
```

Spotify (and the Radio Svizzera Classica app "Swiss Classic") don't seem to actually pause -- they are either `Playing` or `Stopped`.

BTW, the latest version also adds the client name (e.g. "Joe's iPhone") and the service name (e.g. "Den Speakers").

**Comment by phelbas** [comment]
Thanks I downloaded the dev branch and built It

Happy to say the dev branch sends play/pause signals and does allow query of player state Playing/Paused so I can fully reflect the pause/play state into the button on the UI, and update the crawling progress bar.

Is there a way to query the version of shairport-sync to know it is >= this version.
perhaps presence of client name and service name?

(ProgressString doesn't show any change over time but wasn't using that)

**Comment by mikebrady** [comment]
Great stuff.

You can pick up the version string in the D-Bus interface, but it's still at 4.1. Anyway, parsing out the Version, Revision and MinorRevision is a bit of a pain.

I agree it might be easier to check for the existence of the ServiceName property. If it's there, then Shairport Sync has that Play/Pause fix.

Thanks to your work, I've been looking more carefully at the metadata. In particular I've been trying to find a way of telling if what's playing has valid durations. For example, if it's a radio station, it's got no valid length, so a progress bar would be inappropriate. If I find something, I'll post it here.

**Comment by phelbas** [comment]
@mikebrady I've pushed an update to my repo that uses your patches for desktop mode, anti-aliasing (sorry I hand applied them). It also shows the source/sink, rounds the corners on the art, fixes the drop shadow, changes the font slightly and fixes some other mode bugs and unnecessary flashing.

**Comment by mikebrady** [comment]
That looks great! I'll pull the updates over the weekend and give it a try.

The frame rate (usually 44,100 frames per second) and bit depth (32, 24, 16 bits) being used on the output device can certainly be added to the D-Bus interface.

Regarding the source, it's either a Realtime stream (ALAC, 44.1k 16-bit stereo) or a Buffered (256k AAC, 44.1k 24-bit floating point stereo) for "Buffered" streams, and that's about it, so not particularly interesting. FWIW, the type of stream -- Realtime or Buffered -- can be added to the D-Bus interface if it's of interest. (?)

**Comment by mikebrady** [comment]
Nice!

<img width="800" alt="image" src="https://user-images.githubusercontent.com/4265913/201491752-10431c37-fe66-470e-956a-ea69107d191a.png">

**Comment by phelbas** [comment]
Thanks for trying it quickly. I've got a couple of things on my mind to touch up like removing the display of source and destination when disconnected (it lingers while the rest of the widgets vanish), using older shairport-sync versions gracefully and adding a volume slider plus I'll add the bit rate indicator too.
Also fixing the display of streams with no length.

**Comment by phelbas** [comment]
added a volume control which interestingly isn't without minor issues. for example, changing volume via dbus does not change the slider position on the IOS device, but does change the volume ok.
while changing the slider on the IOS device is reflected through correctly. 

also discovered that switching Music from pi, to Phone speaker, and back to pi again, works for a bunch (6 or so) of cycles, then wedges shairport-sync daemon. I need to look at the logs to see if I can see why.

**Comment by phelbas** [comment]
hmm odd. works for me. Line 285 sets the volume via dbus (-30 to 0.0) and line 259 in _tickEvent fetches the current volume, every 5 seconds, and sets the slider to whatever it gets back?

**Comment by mikebrady** [comment]
Thanks. I'll play around with it and see what's happening here.

**Comment by mikebrady** [comment]
Working now! Don't know what was happening.

**Comment by mikebrady** [comment]
So I had a good look at this and played around -- it's really improving!

I made some changes to make the volume control more responsive along the following lines:
* When you send a new volume control request to SPS over D-Bus, the volume actually set in the player comes back immediately in a `PropertiesChanged` message, so you don't actually have to check for it periodically.
* If the volume is changed on the player, it also comes through on a `PropertiesChanged` message.

To deal with both of these situations, when a `AirplayVolume` message comes in, one can block signals from the slider (to stop it feeding back new values), set its value to the new volume and then unblock signals. Also I added some code to get the volume from SPS to give the slider an initial value.

* I added some code to check for the optional presence of that `songdatakind` token, and if it is there, so set a flag and make stuff visible or invisible as necessary.

Happy to send a PR if you'd like it.

Also, a quick question: is there a nice way to prevent it from closing the window when the player stops?

**Comment by mikebrady** [comment]
Looking nice. This is a radio station app RAI "Le Storie Della Radio". Haven't got these the logic of the FF/RW buttons right yet...


<img width="800" alt="image" src="https://user-images.githubusercontent.com/4265913/201705319-e5eaf113-a3b4-4584-8d17-928eabf347fa.png">

**Comment by phelbas** [comment]
thanks I've not changed anything so if you do a pull request I'll just apply it?

BTW an interesting thing is that on IOS shairpoint-sync shows up as a speakers target, but on OSX, it shows up as a whatever this icon is (next to the label living room), but some Ikea airplay speakers show up with the correct speaker Icon so OSX knows how to label speakers..

<img width="292" alt="Screen Shot 2022-11-15 at 7 57 40 am" src="https://user-images.githubusercontent.com/67846089/201764072-977c644f-9a05-4440-a5f6-7d48bf47619f.png">

**Comment by mikebrady** [comment]
> thanks I've not changed anything so if you do a pull request I'll just apply it?

I'm afraid that since I suggested a pull request, I went a bit crazy with the code. There are many superficial changes, so there's a lot to take in, and you mightn't like it. What I've done instead is pushed everything to [my own fork](https://github.com/mikebrady/shairport-display-qt/tree/master) of your repository. If you wished to try it separately and see if you like it, then I'd be very happy to make a PR to your repository.

Briefly, the things I went a bit crazy with are as follows:

1. In the `shairport-display.qt`, used a 0 to -300 slider to better match the 0.0 to -30.0 range of Airplay volume and simplify the arithmetic.
2. Blocked feedback from the volume slider when it is being updated with new values. This stops the slider going crazy when it is being dragged, though it is still a little jumpy. I haven't tried it on a touch interface.
3. Used the `PropertiesChanged` message from D-Bus to manage volume changes and more.
4. Used the `sps:songdatakind` metadata token, when available, to control visibility of the progress bar and the Play/Rewind/FastForward buttons. So far, it works properly for all the different sources I've tried.
5. Refactored some code to remove redundancies. And renamed the method that handles property changes to better reflect its role.
6. Ran the code through the [black](https://medium.com/gousto-engineering-techbrunch/automate-python-code-formatting-with-black-and-pre-commit-ebc69dcc5e03) python auto-formatter and corrected issues noticed by the QC Creator python linter. Unfortunately, this is what has made the most visible (but mostly superficial) differences.
7. Removed some packages that are no longer needed.

So, apologies for messing this up a bit. OTOH, the code does seem to run well...

> ... so OSX knows how to label speakers...

Yeah, it would be nice to pick that up, but haven't found it yet...

<img width="800" alt="image" src="https://user-images.githubusercontent.com/4265913/201898502-266f2d7f-04a8-4ce0-9c3f-86357bb1b7e7.png">

**Comment by phelbas** [comment]
seems all good. It needed to be re-written even before I looked at it, and I'm not a python programmer.
If you're fine with taking it forward then I'll remove my repo as it does everything I need the only last wish is to show the current bit rate so switching Apple Music preferences between the lossy and lossless options may actually be reflected in the display. 

<img width="456" alt="Screen Shot 2022-11-15 at 10 31 14 pm" src="https://user-images.githubusercontent.com/67846089/201909340-e8db654a-15b1-44ce-90f5-eb616a110e5e.png">

**Comment by mikebrady** [comment]
Thanks. I don't wish to "rain on your parade", as it were; you've brought it forward a great deal. In fact, I was actually thinking of removing my repo, but if you're sure, I'll keep mine up and you can remove yours. 

I haven't found any metadata (yet) that might indicate the original encoding (HQ, Hi-Res, etc.). Unfortunately, as you know, it's all transcoded to AAC or ALAC anyway, before it is sent to Shairport Sync. But I'll keep an eye out for it.

**Comment by mikebrady** [comment]
Also, is there an example of that horizontal pixel shifting anywhere?

**Comment by phelbas** [comment]
yeah https://stackoverflow.com/questions/46505130/creating-a-marquee-effect-in-pyside
